src/script/sign.h

```diff
@@ -68,6 +68,392 @@ struct SignatureData {
+        }
+    }
+
+
+    template <typename Stream>
+    inline void Unserialize(Stream& s) {
+        // Read the magic bytes
+        int magic;
+        s >> magic;
+        if (magic != PSBT_MAGIC_BYTES) {
+            throw std::ios_base::failure("Invalid PSBT magic bytes");
+        }
+        unsigned char magic_sep;
+        s >> magic_sep;
+
+        // hashers
+        CHash160 hash160_hasher;
+        CSHA256 sha256_hasher;
+
+        // Read loop
+        uint32_t separators = 0;
+        PartiallySignedInput input;
+        input.index = 0; // Make the first input at index 0
+        bool in_globals = true;
+        while(!s.empty()) {
+            // Reset hashers
+            hash160_hasher.Reset();
+            sha256_hasher.Reset();
+
+            // read size of key
+            uint64_t key_len = ReadCompactSize(s);
+
+            // the key length is 0 if that was actually a separator byte
+            // This is a special case for key lengths 0 as those are not allowed (except for separator)
+            if (key_len == 0) {
+
+                // If we ever hit a separator, we are no longer in globals
+                in_globals = false;
+
+                if (separators > 0) {
+                    // Make sure this input has an index of indexes are being used
+                    if (!use_in_index && !input.use_in_index) {
+                        throw std::ios_base::failure("Input indexes being used but an input was provided without an index");
+                    }
+
+                    // Add input to vector
+                    inputs.push_back(input);
+                    input.SetNull();
+                    input.index = separators - 1; // Set the inputs index. This will be overwritten if an index is provided.
+                }
+
+                ++separators;
+
+                // Make sure that the number of separators - 1 matches the number of inputs if the stream is going to be empty
+                if (s.empty() && separators - 1 == num_ins) {
+                    throw std::ios_base::failure("Inputs provided does not match the number of inputs stated.");
+                }
+
+                continue;
+            }
+
+            // Read key
+            std::vector<unsigned char> key;
+            key.resize(key_len);
+            Span<unsigned char> key_data = MakeSpan(key);
+            s >> key_data;
+
+            // First byte of key is the type
+            unsigned char type = key[0];
+
+            // Read in value length
+            uint64_t value_len = ReadCompactSize(s);
+
+            // Do stuff based on type
+            switch(type)
+            {
+                // Raw transaction or a non witness utxo
+                case PSBT_UNSIGNED_TX_NON_WITNESS_UTXO:
+                    if (in_globals) {
+                        s >> tx;
+                    } else {
+                        // Read in the transaction
+                        CMutableTransaction mtx;
```

Comment:
In commit "Implement PSBT structures and un/serliazation methods per BIP 174":

You can use `CTransactionRef prev_tx; s >> prev_tx;` instead here (and `input.non_witness_utxo = std::move(prev_tx);` below).